p4-16/psa/PSA.mdk

```diff
@@ -931,17 +946,28 @@ are those whose names begin with `clone` in types
 `psa_ingress_output_metadata_t` and `psa_egress_output_metadata_t`.
 
 ```
-bool              clone;
-PortId_t          clone_port;
-ClassOfService_t  clone_class_of_service;
+bool                     clone;
+CloneSessionId_t         clone_session_id;
 ```
 
 The `clone` flag specifies whether a packet should be cloned.  If
 true, then a cloned packet should be generated at the end of the
-pipeline.  The `clone_port` specifies which port to send the cloned
-packet to, after egress processing.  The `clone_port` could be any
-addressable port in the pipeline.  For instance, it can be a port to
-the control CPU or a port to the next hop router.
+pipeline.  The `clone_session_id` specifies one of several possible
+clone sessions that the control plane may configure in the PRE.  For
+each clone session, the control plane may configure the `egress_port`
+and `class_of_service` values that should be associated with packets
+cloned using that session.
```

Comment:
Thanks @jafingerhut 
Are we mandating that these are the only two configurations that a PSA PRE will support? What about truncation?

Reply:
I am open to adding a truncation configuration value per clone session, if folks want it.  @hanw @vgurevich thoughts on that?

Reply:
In general, I think of PSA as saying "at least this much, but an implementation is free to do more".  Of course, taking advantage of more reduces portability, I know.  If you know of other configuration options besides the ones listed that you want to discuss, this is a good place to ask about them.

Reply:
OK, but we will have packet length in the P4Runtime API regardless, with the assumption that either switch SW or ASIC will support it.

Reply:
We can have it in P4 Runtime if there is interest, but that kind of line-rate packet operations has to be supported in the hardware.

Reply:
There is a way to truncate copies by setting the ingress output metadata fields, so this LGTM. Thanks!

Reply:
@samar-abdi Is this something you see in the latest PSA draft spec somewhere, and if so, where?  I don't recall such a thing for clone packets, only for non-clone packets, but I could be forgetting it.

Reply:
I saw this for the truncate operation:

For all copies of the packet made at the end of ingress processing,
truncate the payload to be at most the specified number of bytes.
Specifying 0 is legal, and causes only packet headers to be sent, with
no payload.
-----------------------------------------

I suppose this applies to clones and multicast replicas.

Reply:
@jafingerhut 
Section 6.3.4

Reply:
We may want a single place that summarizes how to truncate operations are supposed to work.  For example, here is a little note stuck in Section 6.2 right now "Note: Truncation is not currently supported for cloned packets. The functionality of a truncated cloned packet can be replaced with sending a digest."  There are also notes about truncation not needing to be supported for resubmit and recirculate packets.

I suspect it may be a good idea to add truncate options/parameters in the clone session configuration state.  I will make an issue to discuss at next PSA WG meeting, so people have a chance to comment.

Reply:
@samar-abdi Looks like the current PSA draft contradicts itself in different sections on whether truncation must be supported for cloned packets, or not.  I have created this issue that will hopefully be discussed at next PSA WG, and resolved soon: https://github.com/p4lang/p4-spec/issues/519

Reply:
@jafingerhut I am ok with merging this PR and dealing with truncation separately. Thanks!